anaconda_anon_usage/heartbeat.py

```diff
@@ -77,39 +85,100 @@ def attempt_heartbeat(channel=None, path=None, wait=False):
                 break
         else:
             _print("No valid heartbeat channel")
-            _print(line, standalone=True)
             return
         url = urljoin(base, channel or "main") + "/"
     url = urljoin(url, path or HEARTBEAT_PATH)
 
     _print("Heartbeat url: %s", url)
+    if prefix:
+        Context.checked_prefix = prefix
+        _print("Prefix: %s", prefix)
     _print("User agent: %s", context.user_agent)
-    if DRY_RUN:
+
+    if dry_run:
         _print("Dry run selected, not sending heartbeat.")
-    else:
-        session = get_session(url)
-        t = Thread(target=_ping, args=(session, url, wait), daemon=True)
-        t.start()
-        _print("%saiting for response", "W" if wait else "Not w")
-        t.join(timeout=None if wait else 0.1)
-    _print(line, standalone=True)
+        return
+
+    # Build and configure the session object
+    timeout = TIMEOUT / ATTEMPTS
+    context.remote_max_retries = ATTEMPTS - 1
+    # No backoff between attempts
+    context.remote_backoff_factor = 0
+    session = get_session(url)
+
+    # Run in the background so we can proceed with the rest of the
+    # activation tasks while the request fires. The process will wait
+    # to terminate until the thread is complete.
+    t = Thread(target=_ping, args=(session, url, timeout), daemon=False)
+    t.start()
+    if STANDALONE:
+        t.join()
 
 
 def main():
     global VERBOSE
-    global DRY_RUN
     global STANDALONE
-    p = argparse.ArgumentParser()
-    p.add_argument("-c", "--channel", default=None)
-    p.add_argument("-p", "--path", default=None)
-    p.add_argument("-d", "--dry-run", action="store_true")
-    p.add_argument("-q", "--quiet", action="store_true")
-    p.add_argument("-w", "--wait", action="store_true")
-    args = p.parse_args()
     STANDALONE = True
-    VERBOSE = not args.quiet
-    DRY_RUN = args.dry_run
-    attempt_heartbeat(args.channel, args.path, args.wait)
+    VERBOSE = "--quiet" not in sys.argv and "-q" not in sys.argv
+
+    line = "-----------------------------"
+    _print(line)
+    _print("anaconda-anon-usage heartbeat")
+    _print(line)
+
+    def environment_path(s):
```

Comment:
TIL that the `type` argument in argparse arguments is a callable:
https://stackoverflow.com/a/37472037